src/analysis/enums.rs

```diff
@@ -0,0 +1,143 @@
+        let type_ = library
+            .type_(self.type_id)
+            .maybe_ref()
+            .unwrap_or_else(|| panic!("{} is not an enumeration.", self.full_name));
+        type_
+    }
+}
+
+pub fn new(env: &Env, obj: &GObject, imports: &mut Imports) -> Option<Info> {
+    info!("Analyzing enumeration {}", obj.name);
+
+    if !obj.status.need_generate() {
+        return None;
+    }
+
+    if !obj
+        .type_id
+        .map_or(false, |tid| tid.ns_id == namespaces::MAIN)
+    {
+        return None;
+    }
+
+    let enumeration_tid = env.library.find_type(0, &obj.name)?;
+    let type_ = env.type_(enumeration_tid);
+    let enumeration: &library::Enumeration = type_.maybe_ref()?;
+
+    let name = split_namespace_name(&obj.name).1;
+
+    // Mark the type as available within the enum namespace:
+    imports.add_defined(&format!("crate::{}", name));
+
+    let has_get_quark = enumeration.error_domain.is_some();
+    if has_get_quark {
+        imports.add("glib::Quark");
+        imports.add("glib::error::ErrorDomain");
+    }
+
+    let has_get_type = enumeration.glib_get_type.is_some();
+    if has_get_type {
+        imports.add("glib::Type");
+        imports.add("glib::StaticType");
+        imports.add("glib::value::SetValue");
+        imports.add("glib::value::FromValue");
+        imports.add("glib::value::FromValueOptional");
+    }
+
+    if obj.generate_display_trait {
+        imports.add("std::fmt");
+    }
+
+    let mut functions = functions::analyze(
+        env,
+        &enumeration.functions,
+        enumeration_tid,
+        false,
+        false,
+        obj,
+        imports,
+        None,
+        None,
+    );
+
+    // Gir does not currently mark the first parameter of associated enum functions -
+    // that are identical to its enum type - as instance parameter since most languages
+    // do not support this.
+    for f in &mut functions {
+        if f.parameters.c_parameters.is_empty() {
+            continue;
+        }
+
+        let first_param = &mut f.parameters.c_parameters[0];
+
+        if first_param.typ == enumeration_tid {
+            first_param.instance_parameter = true;
+
+            let t = f
+                .parameters
+                .transformations
+                .iter_mut()
+                .find(|t| t.ind_c == 0)
+                .unwrap();
+
+            if let TransformationType::ToGlibScalar { name, .. } = &mut t.transformation_type {
+                *name = "self".to_owned();
+            } else {
+                panic!(
+                    "Enum function instance param must be passed as scalar, not {:?}",
+                    t.transformation_type
+                );
+            }
+        }
+    }
+
+    // Enum to_string functions provide static strings and need special error handling.
+    if let Some(to_string) = functions
+        .iter_mut()
+        .find(|f| f.name == "to_string" && f.status.need_generate())
```

Comment:
One thing we might want to consider here also is `get_name()` functions with the correct signature. That seems to be a common pattern too for enums (and flags? not sure)

Reply:
```
xmlstarlet sel -t -v "//_:bitfield/_:function[contains(@name, \"get_name\")]/@c:identifier" gir-files/*.gir
gst_message_type_get_name
gst_stream_type_get_name
```
On bitfield, and these on enumerations:

```
xmlstarlet sel -t -v "//_:enumeration/_:function[contains(@name, \"get_name\")]/@c:identifier" gir-files/*.gir
gst_debug_level_get_name
gst_event_type_get_name
gst_format_get_name
gst_pad_mode_get_name
gst_query_type_get_name
gst_state_change_get_namegst_install_plugins_return_get_namegst_player_color_balance_type_get_name
gst_player_error_get_name
gst_player_state_get_name
```
(not sure what xmlstarlet is doing to the newlines...)

Afaik they're all marked nullable though...

Reply:
None of those should be `nullable` (in the annotations and .gir file), but yes it would need overrides in `Gir.toml` for now.

Reply:
`get_name` support is added and refactored quite a bit, but someone invested in the C code should check and add `nullable = false` in `Gir.toml` for every affected function. I've disabled the aforementioned `ret.nullable = false` assumption in `special_functions` for enumerations/bitfields.

Reply:
You mean now everything is simply goes through the normal nullable code? That's good, this special casing was a bit problematic :)

If you give me a list of functions where you're unsure, I can tell you if they're nullable or not, or if they need special handling (-> manual implementation). Would that help?

Reply:
> You mean now everything is simply goes through the normal nullable code? That's good, this special casing was a bit problematic :)

Nope, just enumerations and bitfields; with prior code `to_string` on `GLSLVersion`/`GLSLProfile` would otherwise be generated as non-nullable despite containing explicit `nullable="1"` in GstGL.gir. `GLAPI`/`GLPlatform` return non-NULL and have now gotten an explicit override as well.

> If you give me a list of functions where you're unsure, I can tell you if they're nullable or not, or if they need special handling (-> manual implementation). Would that help?

The aforementioned `_get_name` functions extracted with `xmlstarlet`. I'm not unsure, it's just a time-consuming process to find and/or override `[object.function.return] nullable = false`. Perhaps we should wait until `trust_return_value_nullability` can be enabled? Otherwise, here's the full list, including everything outside of flags/enums as well. Note that it only helps if you set up a commit adding/overriding `nullable = false` in Gir directly :grimacing: 

```console
test@gstreamer-rs $ xmlstarlet sel -t -v "//_:function[contains(@name, \"get_name\")]/@c:identifier" gir-files/*.gir
gst_debug_level_get_name
gst_element_state_change_return_get_name
gst_element_state_get_name
gst_event_type_get_name
gst_format_get_name
gst_message_type_get_name
gst_pad_link_get_name
gst_pad_mode_get_name
gst_query_type_get_name
gst_state_change_get_name
gst_stream_type_get_name
gst_debug_level_get_name
gst_event_type_get_name
gst_flow_get_name
gst_format_get_name
gst_message_type_get_name
gst_pad_mode_get_name
gst_query_type_get_name
gst_state_change_get_name
gst_stream_type_get_namegst_install_plugins_return_get_name
gst_install_plugins_return_get_namegst_player_color_balance_type_get_name
gst_player_error_get_name
gst_player_state_get_name
gst_player_color_balance_type_get_name
gst_player_error_get_name
gst_player_state_get_name
```

Reply:
> Nope, just enumerations and bitfields; with prior code to_string on GLSLVersion/GLSLProfile would otherwise be generated as non-nullable despite containing explicit nullable="1" in GstGL.gir.

Yeah those are indeed properly nullable. Annoying :)

> Otherwise, here's the full list, including everything outside of flags/enums as well. Note that it only helps if you set up a commit adding/overriding nullable = false in Gir directly 

If you put that list into the GStreamer MR as a discussion (so we don't forget), I'll add overrides (as needed) for all of them to the `Gir.toml`